### PR TITLE
connect:diagnose and connect:mapping:diagnose

### PR DESCRIPTION
**Pull request by gulopine**
In retrospect, I probably should've put the two commands in separate PRs, but the mapping one is pretty much just a copy/paste of the connection command, but with mapping-specific bits copied from our other mapping commands.

### CODE REVIEW

File: commands/connect/diagnose.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      return yield api.request(context, 'GET', url)
+    }))
+
+    cli.log() // Blank line to separate each section
+    cli.styledHeader(`Connection: ${connection.name}`)
+    displayResults(results.json)
+
+    for (let objectName in results.json.mappings) {
+      cli.log() // Blank line to separate each section
+      cli.styledHeader(`Mapping: ${objectName}`)
```
Comment (alouie-sfdc) [inline commands/connect/diagnose.js]: It would be nice to de-emphasize the `Mapping:` part of this string so that the mapping name stands out more.

Comment (alouie-sfdc) [inline commands/connect/diagnose.js]: Same with `Connection:` in the previous block.

File: commands/connect/mapping-diagnose.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  ],
+  needsApp: true,
+  needsAuth: true,
+  run: cli.command(co.wrap(function * (context, heroku) {
+    context.region = regions.determineRegion(context)
+    let connection = yield api.withConnection(context, heroku)
+    let mapping = yield api.withMapping(connection, context.args.mapping)
+    let results = yield cli.action('Diagnosing mapping', co(function * () {
+      let url = '/api/v3/mappings/' + mapping.id + '/validations'
+      cli.debug(url)
```
Comment (alouie-sfdc) [inline commands/connect/mapping-diagnose.js]: Accidentally left this in?

Comment (gulopine) [inline commands/connect/mapping-diagnose.js]: doh! indeed.

File: commands/connect/diagnose.js (hunk 1 above)
Comment (gulopine) [inline commands/connect/diagnose.js]: Hrm. We don't really have any good options for that within `styledHeader`, so we have a couple options.

One is to roll our own heading, and just not use `styledHeader` at all. That would give us full flexibility, but we'd start to drift away from other CLI plugins by doing so.

The other option is to remove "Mapping:" altogether and just display the name of each mapping. I was worried there'd be confusion between the connection and the mappings if we don't display that identifier, but maybe it's okay?

screen shot 2017-06-27 at 6 04 48 pm

Comment (alouie-sfdc) [inline commands/connect/diagnose.js]: No problem, that was just a minor comment. Looking at the screenshot, I'm liking it without "Mapping:". Can we keep the text that says "Connection:", though? I think that would address your concern about distinguishing between connection info and mapping info.

File: commands/connect/diagnose.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+const api = require('../../lib/connect/api.js')
+const regions = require('../../lib/connect/regions.js')
+const cli = require('heroku-cli-util')
+const co = require('co')
+
+function displayResults (results) {
+  results.errors.forEach(displayResult('red'))
+  results.warnings.forEach(displayResult('yellow'))
+  results.passes.forEach(displayResult('green', false))
+}
```
Comment (alouie-sfdc) [inline commands/connect/diagnose.js]: Love how easy this method is to read.

File: lib/connect/api.js [hunk 4]
```diff
@@ -136,7 +136,6 @@ let withUserConnections = exports.withUserConnections = co.wrap(function * (cont
 exports.withConnection = function (context, heroku) {
   return co(function * () {
     let connections = yield withUserConnections(context, context.app, context.flags, true, heroku)
-    console.log('in withConnection:', connections)
```
Comment (sigmavirus24) [inline lib/connect/api.js]: Oops. Thanks for catching this. I must have forgotten this when I was debugging something. :tada:

**Comment by alouie-sfdc** [comment]
👍

**Pull request merged**